File: leetcode/easy/530-min-abs-diff-bst/easy.cpp
/**
 * @file easy.cpp
 * @author Vortexx2
 * @brief Problem 783 - Minimum Distance Between BST Nodes
 * @date 25-08-2021
 * 
 * Runtime - 0 ms O(n)
 * Memory Usage - 9.7 MB O(n)
 */
#include <algorithm>
#include <climits>
#include <iostream>
#include <vector>

using namespace std;

struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode() : val(0), left(nullptr), right(nullptr) {}
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
  TreeNode(int x, TreeNode *left, TreeNode *right)
      : val(x), left(left), right(right) {}
};

class Solution {
 public:
  vector<int> inorder;
  int minDiffInBST(TreeNode *root) {
    helper(root);
    int minDiff = INT_MAX;
    for (int i = 1; i < inorder.size(); i++)
      minDiff = min(minDiff, inorder[i] - inorder[i - 1]);

    return minDiff;
  }

  void helper(TreeNode *root) {
    if (!root) return;
    helper(root->left);
    inorder.push_back(root->val);
    helper(root->right);
  }
};

File: leetcode/easy/530-min-abs-diff-bst/easy.py
"""
  @author Vortexx2
  Problem 783 - Minimum Distance in BST

  Runtime - 32 ms
  Memory Usage - 14.1 MB
"""
# Definition for a binary tree node.
class TreeNode:
  def __init__(self, val=0, left=None, right=None):
    self.val = val
    self.left = left
    self.right = right


class Solution:
  def minDiffInBST(self, root: Optional[TreeNode]) -> int:
    inorder = []

    def helper(root: TreeNode) -> None:
      if not root:
        return
      helper(root.left)
      inorder.append(root.val)
      helper(root.right)

    helper(root)

    res = float('inf')
    for i in range(1, len(inorder)):
      res = min(res, inorder[i] - inorder[i - 1])

    return res

